tls: Update cipher-list defaults, add command-line switches

jasnell opened the pull request:
1. Pull over the io.js default cipher list updates
2. Adds new `--cipher-list` and `--enable-legacy-cipher-list` command line switches. 
   `--cipher-list` allows a new default cipher list to be specified  
   `--enable-legacy-cipher-list` tells node to use specific older cipher lists
3. Adds new `NODE_CIPHER_LIST` and `NODE_LEGACY_CIPHER_LIST`
   environment variables as alternatives to the command line switches
4. Adds a new `getLegacyCiphers` method to the TLS module
5. A test case is included.

jasnell committed:
tls: more secure defaults

Port of io.js commit: https://github.com/iojs/io.js/commit/77f35861d0217273b9e478f5d35bd7d8e47

Original commit message:

This updates the default cipher suite to an more secure list, which
prefers strong ciphers with Forward Secrecy. Additionally, it enables
`honorCipherOrder` by default.

Noteable effect of this change is that the insecure RC4 ciphers are
disabled and that Chrome negotiates a more secure ECDHE cipher.

Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Fedor Indutny <fedor@indutny.com>
PR-URL: https://github.com/iojs/io.js/pull/826

jasnell commented:
(This is just for master... separate PR's will be coming for backporting changes into
v.0.12.x and v0.10.x)

/cc @joyent/node-coreteam

jasnell commented:
There will be just a couple more updates on this coming soon.. just a few minor late breaking tweaks...

jasnell committed:
tls: command-line switch and envar cipher-list override

Add command line switches and environment variables to override
the default cipher suite in tls.js

`--cipher-list` and `NODE_CIPHER_LIST` can be used to completely
override the default cipher list with a given value.

`--enable-legacy-cipher-list` and `NODE_LEGACY_CIPHER_LIST` can
be used to reset the default cipher list back to a known legacy
value shipped in prior Node.js releases

A new `getLegacyCiphers` method on the tis module allows
programmatic access to the old cipher list defaults.

trevnorris commented on src/node.cc:
```diff
@@ -2929,6 +2929,12 @@ static void PrintHelp() {
+         "                       set to v0.10.38 to use the v0.10.38 list,\n"
+         "                       set to v0.10.39 to use the v0.10.39 list.\n"
+         "                       set to v0.12.2 to use the v0.12.2 list.\n"
+         "                       set to v0.12.3 to use the v0.12.3 list.\n"
```
I think it would be enough to simply list possible options. users could probably figure out from context that each one corresponds to a given version.

jasnell replied:
;)  one hopes, yes.
On Apr 3, 2015 1:25 PM, "Trevor Norris" <notifications@github.com> wrote:

> In src/node.cc
> <https://github.com/joyent/node/pull/14383#discussion_r27753373>:
>
> > @@ -2929,6 +2929,12 @@ static void PrintHelp() {
> >  #endif
> >           "  --enable-ssl2        enable ssl2\n"
> >           "  --enable-ssl3        enable ssl3\n"
> > +         "  --cipher-list=val    specify the default TLS cipher list\n"
> > +         "  --enable-legacy-cipher-list=val \n"
> > +         "                       set to v0.10.38 to use the v0.10.38 list,\n"
> > +         "                       set to v0.10.39 to use the v0.10.39 list.\n"
> > +         "                       set to v0.12.2 to use the v0.12.2 list.\n"
> > +         "                       set to v0.12.3 to use the v0.12.3 list.\n"
>
> I think it would be enough to simply list possible options. users could
> probably figure out from context that each one corresponds to a given
> version.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/pull/14383/files#r27753373>.
>

trevnorris commented on src/node.h:
```diff
@@ -223,6 +223,18 @@ NODE_EXTERN void RunAtExit(Environment* env);
   }                                                                           \
   while (0)
 
+#define NODE_DEFINE_STRING_CONSTANT(target, constant)                         \
```
instead I would recommend that the `Isolate` be the first passed argument.

trevnorris commented on src/node_crypto.h:
```diff
@@ -59,10 +60,51 @@
+                                 "ECDHE-RSA-AES128-SHA256:"                   \
+                                 "DHE-RSA-AES128-SHA256:"                     \
+                                 "HIGH:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:" \
+                                 "!PSK:!SRP:!CAMELLIA"
```
alignment issues on the ending `/`. Check that all the `DEFAULT_CIPHER_LIST_*` are aligned at the end the same.

trevnorris commented:
Few small things, but looking good.

jasnell committed:
tls: make --enable-legacy-cipher-list less verbose

Based on commit feedback, make the PrintHelp for
--enable-legacy-cipher-list less verbose.

jasnell committed:
tls: pass in isolate with define string constant + style nits

Per the commit feedback, fix up style nits and pass in the
isolate with the NODE_DEFINE_STRING_CONSTANT macro.

mdawsonibm commented:
Reviewed for consistency with 0.12 and 0.10 patches looks good.  Differences between them also make sense.  This looks like changes from original io.js patch plus the changes for 0.10.X and 0.12.X.  Will need to squash and ensure final commit comment is less than 50 char limit but otherwise lgtm.

misterdjules commented on src/node.cc:
```diff
@@ -3040,6 +3047,20 @@ static void ParseArgs(int* argc,
     } else if (strcmp(arg, "--v8-options") == 0) {
       new_v8_argv[new_v8_argc] = "--help";
       new_v8_argc += 1;
+    } else if (strncmp(arg, "--cipher-list=", 14) == 0) {
```
Could this be done in `src/node.js` instead of in `src/node.cc`? It seems it would just be less error prone and shorter.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -0,0 +1,69 @@
+
+  switch(useswitch) {
+    case 1:
+      // Test --cipher-test
```
Very minor, but I guess this should read `// Test --cipher-list`.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -0,0 +1,69 @@
+
+function doTest(checklist, env, useswitch) {
+  var options;
+  if (env && useswitch === 1) {
```
It seems that if `usesswitch === 1`, it means that we want to test `--cipher-list`, so why set `options` in this case?

misterdjules commented on src/node_crypto.cc:
```diff
@@ -4851,6 +4852,26 @@ static void array_push_back(const TypeName* md,
+    args.GetReturnValue().Set(
+      v8::String::NewFromUtf8(args.GetIsolate(), list));
+  } else {
+    args.GetReturnValue().Set(
```
Should a warning be displayed if a given legacy ciphers identifier doesn't exist?

misterdjules commented on test/simple/test-tls-dhe.js:
```diff
@@ -47,6 +47,7 @@ function test(keylen, expectedCipher, cb) {
   var options = {
     key: key,
     cert: cert,
+    ciphers: ciphers,
```
Is it actually needed? I thought that `'DHE-RSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256'` were in the default ciphers list and that this test tested that the server should work with these ciphers out of the box, without using the `ciphers` option.

misterdjules commented:
@jasnell Added some comments, sorry again for taking so long in reviewing this!

misterdjules commented:
@jasnell Running the tests suite in `test/external/ssl-options/test.js` with these changes now, I'll keep you posted about the results.

misterdjules commented:
@jasnell All tests in `test/external/ssl-options/test.js` pass.

jasnell commented:
+1... my time tonight is disappearing quick so I will work on landing these
in the morning. Thank you very much for the additional review Julien. It's
very helpful

On Tue, Apr 7, 2015 at 5:42 PM, Julien Gilli <notifications@github.com>
wrote:

> @jasnell <https://github.com/jasnell> All tests in
> test/external/ssl-options/test.js pass.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/pull/14383#issuecomment-90769236>.
>

jasnell committed:
test: fixing a few nits in the test

typo and unnecessary options init

jasnell commented:
Landed
